### ISSUE TITLE
Memory leak of stream buffers? (on CLISP)

### ISSUE THREAD

**Issue by avodonosov** [cl-plus-ssl/cl-plus-ssl#120]
The `ssl-stream` class [allocates](https://github.com/cl-plus-ssl/cl-plus-ssl/blob/51885381423cfbdbc3fb3b2eef743f08045c6907/src/streams.lisp#L43) input and output buffers, but they are not de-allocated anywhere.

The [`close`](https://github.com/cl-plus-ssl/cl-plus-ssl/blob/51885381423cfbdbc3fb3b2eef743f08045c6907/src/streams.lisp#L71)
 method seems to be the right place to de-allocate the buffers.

In most lisps the buffer is [created](`https://github.com/cl-plus-ssl/cl-plus-ssl/blob/51885381423cfbdbc3fb3b2eef743f08045c6907/src/ffi-buffer.lisp#L8`) with `cffi:make-shareable-byte-vector` what should be garbage collected I think,  but in CLISP the buffer is [created](https://github.com/cl-plus-ssl/cl-plus-ssl/blob/51885381423cfbdbc3fb3b2eef743f08045c6907/src/ffi-buffer-clisp.lisp#L48) with `(cffi-sys:%foreign-alloc size)`

**Comment by svetlyak40wt** [comment]
Interesting. On Lispworks I also experience memory leaks when accessing HTTPS sites via dexador. Dexador uses CL+SSL [like this](https://github.com/fukamachi/dexador/blob/master/src/backend/usocket.lisp#L465-L489).

On Lispworks this version of the `make-buffer` is used:

```
(defun make-buffer (size)
  (cffi:make-shareable-byte-vector size))
```

and `make-shareable-byte-vector` just creates a statically allocated array:

```
(defun make-shareable-byte-vector (size)
  "Create a shareable byte vector."
  #+(or lispworks3 lispworks4 lispworks5.0)
  (sys:in-static-area
    (make-array size :element-type '(unsigned-byte 8)))
  #-(or lispworks3 lispworks4 lispworks5.0)
  (make-array size :element-type '(unsigned-byte 8) :allocation :static))
```

I didn't find in the LW documentation any information about the need for manual deallocation of such arrays.

**Comment by avodonosov** [comment]
Yes, LW docs [say](http://www.lispworks.com/documentation/lw70/LW/html/lw-1272.htm) the arrays created in static area are garbage collected.

How does memory leak manifest itself? Are you sure that's a leak? (memory usage can grow simply because GC hasn't yet collected some of the objects, but it will collect it later)

**Comment by svetlyak40wt** [comment]
I've checked if static vectors are garbage collected. And they do. However, leak is still there.

Here is how to reproduce it:

```lisp
CL-USER 2 > (ql:quickload :dexador)

CL-USER 3 > (room)
 > Generation 7 allocated 34932448
 > Generation 6 allocated 0
 > Generation 5 allocated 0
 > Generation 4 allocated 0
 > Generation 3 allocated 1308696
 > Generation 2 allocated 36477480
 > Generation 1 allocated 24770160
 > Generation 0 allocated 899456

Total allocated 98388240 (0x5DD4910), total size 230957056 (0xDC42000) 

CL-USER 4 > (loop repeat 1000 do (dex:get "https://ya.ru"))

Warning: cl+ssl::input-wait is not implemented for this lisp, but a non-blocking stream is encountered
Warning: cl+ssl::input-wait is not implemented for this lisp, but a non-blocking stream is encountered
Warning: cl+ssl::input-wait is not implemented for this lisp, but a non-blocking stream is encountered
(and 3851147 more such lines)

CL-USER 5 > (gc-all)

CL-USER 6 > (room)
 > Generation 7 allocated 34508528
 > Generation 6 allocated 0
 > Generation 5 allocated 0
 > Generation 4 allocated 0
 > Generation 3 allocated 1639336544
 > Generation 2 allocated 2165248
 > Generation 1 allocated 100272
 > Generation 0 allocated 2593448

Total allocated 1678704040 (0x640EFDA8), total size 1700970496 (0x6562C000) 
```

So, 1580315800 bytes was not freed.

But then I do "Clear Listener" command to remove all these warnings and run GC again:

```lisp
CL-USER 9 > (gc-all)
82303512

CL-USER 10 > (room)
 > Generation 7 allocated 34497368
 > Generation 6 allocated 0
 > Generation 5 allocated 0
 > Generation 4 allocated 0
 > Generation 3 allocated 46369536
 > Generation 2 allocated 1406664
 > Generation 1 allocated 29944
 > Generation 0 allocated 1131168

Total allocated 83434680 (0x4F91CB8), total size 106024960 (0x651D000) 
```

So, all memory was freed. This is not a memory leak in the CL+SSL, but a leak because of the numerous warnings.

Is there a way to get rid of these "Warning: cl+ssl::input-wait is not implemented for this lisp, but a non-blocking stream is encountered"?

**Comment by avodonosov** [comment]
The best way is to implement `input-wait` and `output-wait` functions, as it is done for other lisps: https://github.com/cl-plus-ssl/cl-plus-ssl/blob/a093514267990d5faa0b522e82c41446d39cc151/src/ffi.lisp#L869

**Comment by svetlyak40wt** [comment]
@avodonosov please, see this pull-request with INPUT-WAIT implementation: https://github.com/cl-plus-ssl/cl-plus-ssl/pull/130
